### PR TITLE
docker/install: Fix latest image install on lima

### PR DESCRIPTION
**Pull request by vvoland**
`latest` is not a valid git tag or revision to get the matching systemd unit files.
Look up the exact source git commit from the `org.opencontainers.image.revision` image config label.

### CODE REVIEW

**Comment by crazy-max** [comment]
https://github.com/docker/actions-toolkit/actions/runs/11591165912/job/32270352253?pr=480#step:10:745

```
        wget https://raw.githubusercontent.com/moby/moby/27.3.1/contrib/init/systemd/docker.service         -O /etc/systemd/system/docker.service
        wget https://raw.githubusercontent.com/moby/moby/27.3.1/contrib/init/systemd/docker.socket         -O /etc/systemd/system/docker.socket
```

leads to 404: https://raw.githubusercontent.com/moby/moby/27.3.1/contrib/init/systemd/docker.service         -O /etc/systemd/system/docker.service

Looks commit sha is not there? Guess it should be: https://github.com/moby/moby/blob/41ca978a0a5400cc24b274137efa9f25517fcc0b/contrib/init/systemd/docker.service

**Comment by vvoland** [comment]
Oh right, that URL only worked for tags, not refs.

File: src/docker/install.ts [hunk 1]
```diff
@@ -127,12 +134,22 @@ export class Install {
...
           core.info(`Downloading dockerd from moby/moby-bin:${tag}`);
-          const moby = await HubRepository.build('moby/moby-bin');
           await moby.extractImage(tag, extractFolder);
         } else if (platform == 'darwin') {
-          // On macOS, the docker daemon binary will be downloaded inside the lima VM
+          // On macOS, the docker daemon binary will be downloaded inside the lima VM.
+          // However, we will get the exact git revision from the image config
+          // to get the matching systemd unit files.
+          const manifest = await moby.getPlatformManifest(tag);
+          const config = await moby.getConfig<ImageConfig>(manifest.config.digest);
```
Comment (crazy-max) [inline src/docker/install.ts:146]: I think labels and annotations will only by on the manifest list config

Comment (vvoland) [reply src/docker/install.ts:146]: Looking at `docker buildx imagetools inspect  moby/moby-bin:latest --raw `, index doesn't have these:
```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:c52ff4ad684b729f597eefacf1abc65913dcb3423071b71997cd271690c8a9b7",
      "size": 1630,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:fda20c7387897ee54811fd74b96472c4a454a62f29b5eb595caaeb4e04f01bd5",
      "size": 567,
      "annotations": {
        "vnd.docker.reference.digest": "sha256:c52ff4ad684b729f597eefacf1abc65913dcb3423071b71997cd271690c8a9b7",
        "vnd.docker.reference.type": "attestation-manifest"
      },
      "platform": {
        "architecture": "unknown",
        "os": "unknown"
      }
...
```

Comment (vvoland) [reply src/docker/install.ts:146]: The image config has them though:
```json
docker buildx imagetools inspect  moby/moby-bin:latest@sha256:fe77a2f844e3db621de180b88e0e8c5439dd6802e4e6efcc5371be4a05e252d3 --raw | jq
{
  "architecture": "amd64",
  "config": {
    "Env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    ],
    "WorkingDir": "/",
    "Labels": {
      "org.opencontainers.image.created": "2024-09-20T18:10:20.768Z",
      "org.opencontainers.image.description": "The Moby Project - a collaborative project for the container ecosystem to assemble container-based systems",
      "org.opencontainers.image.licenses": "Apache-2.0",
      "org.opencontainers.image.revision": "41ca978a0a5400cc24b274137efa9f25517fcc0b",
      "org.opencontainers.image.source": "https://github.com/moby/moby",
      "org.opencontainers.image.title": "moby",
      "org.opencontainers.image.url": "https://github.com/moby/moby",
      "org.opencontainers.image.version": "27.3.1"
    }
  },
```

Comment (crazy-max) [reply src/docker/install.ts:146]: Good catch

**Comment by vvoland** [comment]
Drafted for now and added some debug logs. Will undraft once fixed.

**Comment by vvoland** [comment]
Fixed! The issues were:

- Trying to get the config for the darwin platform (there's no Engine build for darwin)
- Casting the `application/vnd.oci.image.config.v1+json` object directly to a `ImageConfig` struct (when it's actually `Image` that has the config embedded as `config` 🤦🏻)

**Review by crazy-max — APPROVED** [review]
LGTM thanks!

File: src/docker/install.ts [hunk 2]
```diff
@@ -127,12 +130,28 @@ export class Install {
...
+          // On macOS, the docker daemon binary will be downloaded inside the lima VM.
+          // However, we will get the exact git revision from the image config
+          // to get the matching systemd unit files.
+          core.info(`Getting git revision from moby/moby-bin:${tag}`);
+
+          // There's no macOS image for moby/moby-bin - a linux daemon is run inside lima.
+          const manifest = await moby.getPlatformManifest(tag, 'linux');
+
+          const config = await moby.getJSONBlob<Image>(manifest.config.digest);
+          core.info(`Config ${JSON.stringify(config.config)}`);
```
Comment (crazy-max) [inline src/docker/install.ts:147]: Either remove this one or put it in debug mode: `core.debug`

Comment (vvoland) [reply src/docker/install.ts]: Ah good point, done

**Review by crazy-max — APPROVED** [review]

**Pull request merged**